trieste/acquisition/rule.py

```diff
@@ -211,31 +212,208 @@ def acquire(
+            ``builder``. This should *maximize* the acquisition function, and must be compatible
+            with the global search space. Defaults to
+            :func:`~trieste.acquisition.optimizer.automatic_optimizer_selector`.
+        """
+        if builder is None:
+            builder = BatchMonteCarloExpectedImprovement(10_000)
+
+        if optimizer is None:
+            optimizer = automatic_optimizer_selector
+
+        if isinstance(builder, SingleModelAcquisitionBuilder):
+            builder = builder.using(OBJECTIVE)
+
+        self._builder: AcquisitionFunctionBuilder = builder
+        self._optimizer = optimizer
+        self._acquisition_function: Optional[AcquisitionFunction] = None
+
+    def __repr__(self) -> str:
+        """"""
+        return f"""AsynchronousOptimization(
+        {self._builder!r},
+        {self._optimizer!r})"""
+
+    def acquire(
+        self,
+        search_space: SP_contra,
+        datasets: Mapping[str, Dataset],
+        models: Mapping[str, ProbabilisticModel],
+    ) -> types.State[AsynchronousRuleState | None, TensorType]:
+        """
+        Constructs a function that, given ``AsynchronousRuleState``,
+        returns a new state object and points to evaluate.
+        The state object contains currently known pending points,
+        that is points that were requested for evaluation,
+        but observation for which was not received yet.
+        To keep them up to date, pending points are compared against the given dataset,
+        and whatever points are in the dataset are deleted.
+
+        Let's suppose we have P pending points. To optimize the acquisition function
+        we call it with batches of size P+1, where first P points are fixed to pending points.
+        Optimization therefore happens over the last point only, which is returned.
+
+        :param search_space: The local acquisition search space for *this step*.
+        :param datasets: The known observer query points and observations.
+        :param models: The models of the specified ``datasets``.
+        :return: A function that constructs the next acquisition state and the recommended query
+            points from the previous acquisition state.
+        """
+        if self._acquisition_function is None:
+            self._acquisition_function = self._builder.prepare_acquisition_function(
+                datasets, models
+            )
+        else:
+            self._acquisition_function = self._builder.update_acquisition_function(
+                self._acquisition_function, datasets, models
+            )
+
+        def state_func(
+            state: AsynchronousRuleState | None,
+        ) -> tuple[AsynchronousRuleState | None, TensorType]:
+            tf.debugging.Assert(self._acquisition_function is not None, [])
+
+            if state is None:
+                state = AsynchronousRuleState(None)
+
+            state = state.subtract_points(datasets[OBJECTIVE].query_points)
```

Comment:
why would we want to keep duplicate points? and why would duplicates arise in the first place?

Reply:
if you have noise, it can make sense to choose to sample twice the same point. In some cases, the maximiser of the acquisition function can be an existing point. 
I think that making sure that we don't remove more points than the newly acquired ones should do the trick. E.g. if you have 3 identical pending points and you get 1 back, it doesn't matter which pending point you're substracting but you should not remove more than one.

Reply:
this is now implemented, but turned out to be very hard to do efficiently, so i suspect current implementation may trigger tf graph evaluation prematurely. If anyone has better ideas on how to implement it, please suggest